civihr_employee_portal/src/Security/PublicFirewall.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+namespace Drupal\civihr_employee_portal\Security;
+
+/**
+ * This class is responsible for checking if anonymous users can access
+ * certain routes
+ */
+class PublicFirewall {
+
+  /**
+   * Anonymous users can access these paths
+   *
+   * @var array
+   */
+  public static $publicRoutes = [
+    '/^welcome-page$/',
+    '/^sites\/default\/files\/logo.jpg$/', // if logo is missing
+    '/^request_new_account\/ajax$/', // from login page
+    '/^user((?!\/register).)*$/', // user* except user/register
+    '/^yoti-connect*/' // yoti login plugin
+  ];
+
+  /**
+   * @param \stdClass $user the user to check access for
+   * @param string $route the route to check
+   *
+   * @return bool
+   */
+  public function canAccess($user, $route) {
+    $isAnonymous = is_null($user) || $user->uid == 0;
+
+    if (!$isAnonymous) {
+      return TRUE;
+    }
+
+    return $this->allowAnonymousAccess($route);
+  }
+
+  /**
+   * Check if route matches against all public routes
+   *
+   * @param string $route
+   * @return bool
+   */
+  private function allowAnonymousAccess($route) {
+    foreach (self::$publicRoutes as $pattern) {
+      if (preg_match($pattern, $route)) {
```

Comment:
given that most URL's are fixed and don't really required a regex, I'm thinking of a little improvement: For fixed routes, we just use a simple string, but for regex, we should prefix it with something (and @, for example):

```php
public static $publicRoutes = [
    'welcome-page',
    'sites/default/files/logo.jpg', // if logo is missing
    'request_new_account/ajax', // from login page
    '@/^user((?!\/register).)*$/', // user* except user/register
    '@/^yoti-connect*/' // yoti login plugin
];
```

Then this method would be something like:
```php
foreach (self::$publicRoutes as $routes) {
  if ($route[0] == '@') {
     // preg_match with the string without the @
  } else {
    // just a normal comparison
  }
}
```

The advantages:
- you don't need the more complex syntax and escaping for fixed URLs
- you avoid unnecessary regex checks